File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.util.*;
import java.io.*;

class Point {
	int x, y;

	Point(int x, int y) {
		this.x = x;
		this.y = y;
	}

	int getDistance(int x, int y) {
		return Math.abs(this.x - x) + Math.abs(this.y - y);
	}
}

public class Main {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;
	static List<Point> home = new ArrayList<>();
	static List<Point> chicken = new ArrayList<>();
	static int N, M;
	static int result = Integer.MAX_VALUE;

	public static void main(String[] args) throws IOException {
		int[][] map;
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		map = new int[N][N];

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
				if (map[i][j] == 1)
					home.add(new Point(i, j));
				else if (map[i][j] == 2)
					chicken.add(new Point(i, j));
			}
		}

		chickChoice(new int[M], 0, 0, 0);

		bw.write(result + "");
		bw.close();
	}

	// M개의 치킨집고르기
	static void chickChoice(int[] path, int start, int depth, int flag) {
		if (depth == M) {
			int sum = 0;
			for (Point h : home) {
				int min = Integer.MAX_VALUE;
				for (int i = 0; i < path.length; i++) {
					min = Math.min(h.getDistance(chicken.get(path[i]).x, chicken.get(path[i]).y), min);
				}
				sum += min;
			}
			result = Math.min(sum, result);
			return;
		}

		for (int i = start; i < chicken.size(); i++) {
			if ((flag & (1 << i)) != 0)
				continue;
			path[depth] = i;
			chickChoice(path, i + 1, depth + 1, flag | 1 << i);
		}
	}
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
import itertools, collections, math, sys

input = lambda: sys.stdin.readline().rstrip()


def get_dist(r1, c1, r2, c2):
    return abs(r1 - r2) + abs(c1 - c2)


N, M = map(int, input().split())


board = [list(map(int, input().split())) for _ in range(N)]

chicken = []
home = []

for i in range(N):
    for j in range(N):
        if board[i][j] == 1:
            home.append((i, j))
        if board[i][j] == 2:
            chicken.append((i, j))

result = float("inf")
for cxy in itertools.combinations(chicken, M):
    total = 0
    for hx, hy in home:
        dist = float("inf")
        for cx, cy in cxy:
            dist = min(dist, get_dist(hx, hy, cx, cy))
        total += dist

    result = min(total, result)

print(result)
